core/src/main/scala/com/yahoo/maha/core/query/QueryGenerator.scala

```diff
@@ -260,18 +260,45 @@ object QueryGeneratorHelper {
   }
 }
 
+
+case class Version(value: Int)
+object Version {
+  val DEFAULT = V0
+  def of(ver: Int): Version = {
+    ver match {
+      case 0 => V0
+      case 1 => V1
+      case 2 => V2
+    }
+  }
+}
+object V0 extends Version(0)
+object V1 extends Version(1)
+object V2 extends Version(2)
+
 class QueryGeneratorRegistry {
-  private[this] var queryGeneratorRegistry : Map[Engine, QueryGenerator[_]] = Map.empty
 
-  def isEngineRegistered(engine: Engine): Boolean = synchronized {
-    queryGeneratorRegistry.contains(engine)
+  private[this] var queryGeneratorRegistry : Map[Engine, Map[Version , QueryGenerator[_]]] = Map.empty
+
+  def isEngineRegistered(engine: Engine, version: Version = Version.DEFAULT): Boolean = synchronized {
```

Comment:
You should not have default value for version, it should be passed in, otherwise you may end up verifying an engine exists for default version and not the version passed into getGenerator.  You should make this method take same params as getGenerator.

Reply:
We can't know which qgen version to use before the engine is decided. 'isEngineRegistered' is called  in the decision process. I have made it an option and removed default so that callers are forced to pass a value.